Refactor of error handling

rbren opened the pull request:
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [x] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below
CHANGELOG: Overhaul of error handling and reporting

This PR puts forward a new philosophy for error handling in OpenHands:
* Errors caused by the agent being stupid are ErrorObservations
  * these the agent can recover from automatically with feedback
* Errors caused by fatal environment problems are Exceptions
  * these stop the agent loop
  * the user must take some action to continue
  * some plumbing involved to get them back to the user, but runtime and agent_controller are both working nicely

Both types show up as messages in the chat window (no more toasts!). Both types of errors MAY include an `error_id`, which triggers a nice, translated message in the FE, and hides the (potentially ugly) details behind a button.

If no `error_id` is included, we fall back to just showing the details, which may be garbled nonsense like a stack trace or JSON from an API.

<img width="280" alt="Screenshot 2024-10-31 at 4 20 18 PM" src="https://github.com/user-attachments/assets/3eb156f4-a71e-4ded-9fd2-073745da42f0">
<img width="264" alt="Screenshot 2024-10-31 at 4 20 28 PM" src="https://github.com/user-attachments/assets/f7a118f8-e2a0-4c6d-8fa9-3a898518c941">

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

Right now, we have three types of errors:
* ErrorObservation: tend to be triggered by the agent doing something stupid, which it can recover from, like sending a non-existent action or making a bad edit
  * the agent can recover from these, so dropping them into the eventstream makes sense
* FatalErrorObservation: tend to be triggered by critical errors like the runtime going down unexpectedly
  * the agent cannot recover from these, but they need to be reported to the user
* normal python exceptions: crop up randomly, and are sometimes converted into one of the above
  * these are unexpected

This PR does away with FatalErrorObservation, and instead always raise exceptions for Errors that are unrelated to the agent doing its business.

ErrorObservations will continue to flow into the event stream, but we will never stop the loop. They will show up in the message window.

Exceptions will _not_ flow into the event stream. Instead, they will stop the loop, and be sent back to the user, where they show up as a message in the chat window.

There are two major error flows that I've tested:
* Wrong API key (currently the agent just hangs on the frontend, with this PR you get a nice message)
* Runtime disconnected (currently we do a decent job here, but this makes it more robust)

I also removed a lot of the `retry` logic in EventStreamRuntime, which is a bit of an anti-pattern, and was hiding a lot of pain (e.g. something that obviously wasn't going to work would just stall the system for 5m because we retried on every Exception)


---
**Link of any specific issues this addresses**

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=ghcr.io/all-hands-ai/runtime:dc02120-nikolaik   --name openhands-app-dc02120   ghcr.io/all-hands-ai/runtime:dc02120
```

enyst commented on openhands/events/stream.py:
```diff
@@ -138,6 +138,10 @@ def add_event(self, event: Event, source: EventSource):
+        if hasattr(event, '_id') and event.id is not None:
+            raise ValueError(
+                'Event already has an ID. It was probably added back to the EventStream from inside a handler, trigging a loop.'
+            )
```
This should work for that case! Although it will fail hard with ValueError, and the FatalObs is in the eventstream, but I think evals don't access the stream directly.

If so, maybe we can add a `cause` to this or something, so that it can be discovered on the other side and reported?

enyst replied:
Actually, I don't know, can you please point out where is it sending the same? I see `send_error` is creating another ErrorObs. 🤔

rbren replied:
Ohh I think you're right actually :/

It's late. Going to switch this to draft

enyst commented on openhands/controller/agent_controller.py:
```diff
@@ -133,21 +133,19 @@ async def update_state_after_step(self):
-            ErrorObservation(f'{message}:{detail}'), EventSource.USER
+            detail += '\nPlease check your credentials. Is your API key correct?'
+        await self.event_stream.async_add_event(
+            ErrorObservation(f'{message}:{detail}'), EventSource.AGENT
```
This method was initially meant only for informing the user in the UI. But it had such a good name, that we felt like it should be used almost like logging errors. 😂

rbren replied:
😂 

TBH I don't think an AgentController should know anything about a UI

enyst replied:
Well, it's not that it knows it, strictly speaking, just the ErrorObs created here has a message intended for the chat box / end user's eyes.

And by this time, it doesn't have anymore the info about the exception, except... it did, when it got it as param for some exception info intended for evals.

enyst replied:
Actually, you're right. How about:
- we don't do at all these UI-intended lines of code here, in the controller, 
- create an ErrorObservation intended for the backend / LLM
- with exception information
- put it in the stream
- let the agent session interpret it and send over whatever info for the UI it makes of it.

enyst commented on openhands/events/observation/error.py:
```diff
@@ -12,6 +12,7 @@ class ErrorObservation(Observation):
     E.g., Linter error after editing a file.
     """
 
+    fatal: bool = False
```
The reason I thought adding a FatalErrorObservation is a good thing, is that we treat ErrorObs too much the same, when there are essential differences, such as: 
- some should be in agent's history, some not;
- some should have a user-friendly message, some should have an LLM-friendly content.

In this case, it seems we can also add:
- some should stop the controller, some should not.

Not distinguishing them well right now, is IMHO a pretty big part of what is confusing in our current error handling. The alternative might be to go back to ErrorObs are in one category, and exceptions in the other... I guess.

rbren replied:
Hmm. I do think it's important to distinguish them (hence the `fatal=True`) but we end up having to check `isinstance(ErrorObs) or isinstance(FatalErrorObs)`, and it seems ripe for someone forgetting to handle one or the other, or for the list of error types to grow indefinitely

In general I want to pare down the number of unique things that could end up in the EventStream, because as we add long-term state, we're going to have to ensure old EventStreams still deserialize properly, and the smaller the footprint the better. 

> some should stop the controller, some should not.
This seems like the only thing `fatal` should really distinguish. I.e. if you hit a `fatal` Error, the event stream stops until there's a user interaction. I see what you mean about one being user-facing, and the other being LLM-facing though...

Interestingly, as I look through a lot of the fatal ones, they seem like things the agent should be able to recover from (e.g. submitting a non-existent action)

Will think more about this

enyst replied:
Aha, yes, good points. Re: deserialization, I just hit something similar.

About this bit: 
> I see what you mean about one being user-facing, and the other being LLM-facing though...

What do you think about adding a "cause" or "error" attr to the ErrorObs, with the underlying exception string, at least, if not more of the underlying exception? (We're already using a "cause" attribute, maybe we could still rename that to "parent", but yep deserialization)

So that the agent has an ErrorObs with that info. I seem to recall having tried it, though, but I didn't like the result, because we may want to be a bit careful to make it very easy for developers to use in all the places where an ErrorObs is created or where that underlying exception is handled. Otherwise it will be forgotten, as you just said. 😅 But that was a while ago, maybe it's worth another look if there is some reasonable way to do it?

enyst replied:
Actually, while the `isinstance()` issue is a huge red flag, that could be taken care of by subclassing...

enyst commented on openhands/controller/agent_controller.py:
```diff
@@ -133,21 +131,16 @@ async def update_state_after_step(self):
+            # it's important to set the state before adding the error event, so that metrics sync properly
+            await self.set_agent_state_to(new_state)
+        await self.event_stream.async_add_event(
+            ErrorObservation(message), EventSource.AGENT
```
Are you sure about changing USER to AGENT? We were just discussing [about this](https://github.com/All-Hands-AI/OpenHands/pull/4387#discussion_r1800341181).

enyst replied:
This might not matter much with tool use, because that one is `role=tool`! Or I hope it stays role tool, anyway. (I'm not sure yet if/how all tool use APIs support the role)

enyst replied:
_Edited to add:_ oh! For the UI we need it AGENT, right? Well... so this is the mix UI/LLM in ErrorObservations hitting. 🤔

We could adjust it in the agent for obs... but if that's what you want to go with, please lets add some comment. We _will_ forget, because some of us reflexively assume that the LLM's way is _the way things should be_. 😂

Maybe that idea linked above, to have in our code an "environment" (or "tool") role, is not too bad.

rbren replied:
Thanks for solving this one!

ngduyanhece commented:
## PR Reviewer Comments

After analyzing the pull request, here are my findings:

**Overall Feedback:**

The minor refactor of the agent controller is a positive step towards improving the codebase's maintainability and clarity. The changes made to handle errors more effectively and the removal of `FatalErrorObservation` in favor of a more general `ErrorObservation` are commendable. However, there are opportunities to enhance the code further, particularly in terms of consistency and error handling.

**Score:** 87/100

**Labels:** Bug fix, Enhancement

**Code Suggestions:**

1. **File:** `openhands/runtime/utils/bash.py`
   - **Suggestion Content:** Ensure that the `run` method consistently returns the same type of observation to avoid type-related issues.
   - **Relevant Line:** `+    def run(self, action: CmdRunAction) -> CmdOutputObservation | ErrorObservation:`
   - **Existing Code:**
     ```python
     def run(self, action: CmdRunAction) -> CmdOutputObservation | FatalErrorObservation:
     ```
   - **Improved Code:**
     ```python
     def run(self, action: CmdRunAction) -> CmdOutputObservation | ErrorObservation:
     ```

2. **File:** `openhands/runtime/utils/edit.py`
   - **Suggestion Content:** Consider using a consistent naming convention for error handling to improve readability and maintainability.
   - **Relevant Line:** `+                return ErrorObservation(...)`
   - **Existing Code:**
     ```python
     return ErrorObservation(
         f'Fatal Runtime in editing: Expected FileWriteObservation, got {type(obs)}: {str(obs)}',
     )
     ```
   - **Improved Code:**
     ```python
     return ErrorObservation(
         f'Error in editing: Expected FileWriteObservation, got {type(obs)}: {str(obs)}',
     )
     ```

3. **File:** `tests/unit/test_agent_controller.py`
   - **Suggestion Content:** Ensure that the test names clearly reflect the functionality being tested for better clarity.
   - **Relevant Line:** `+async def test__react_to_error(mock_agent, mock_event_stream):`
   - **Existing Code:**
     ```python
     async def test_report_error(mock_agent, mock_event_stream):
     ```
   - **Improved Code:**
     ```python
     async def test_react_to_error(mock_agent, mock_event_stream):
     ```

The minor refactor of the agent controller is a positive step! Here are some suggestions:

1. Ensure that the `run` method consistently returns the same type of observation to avoid type-related issues.
2. Consider using a consistent naming convention for error handling to improve readability.
3. Ensure that test names clearly reflect the functionality being tested for better clarity.

ngduyanhece commented on openhands/runtime/utils/bash.py:
```diff
@@ -275,7 +275,7 @@ def _continue_bash(
         return output, exit_code
 
-    def run(self, action: CmdRunAction) -> CmdOutputObservation | FatalErrorObservation:
+    def run(self, action: CmdRunAction) -> CmdOutputObservation | ErrorObservation:
```
**Suggestion:** Ensure consistent return types in the `run` method.

**Existing Code:**
```Python
def run(self, action: CmdRunAction) -> CmdOutputObservation | FatalErrorObservation:
```

**Improved Code:**
```Python
def run(self, action: CmdRunAction) -> CmdOutputObservation | ErrorObservation:
```

**Details:**
Ensure that the `run` method consistently returns the same type of observation to avoid type-related issues.

ngduyanhece commented on openhands/runtime/utils/edit.py:
```diff
@@ -214,8 +213,9 @@ def edit(self, action: FileEditAction) -> Observation:
             if not isinstance(obs, FileWriteObservation):
-                return FatalErrorObservation(
-                    f'Fatal Runtime in editing: Expected FileWriteObservation, got {type(obs)}: {str(obs)}'
+                return ErrorObservation(
```
**Suggestion:** Use consistent naming for error handling to improve readability.

**Existing Code:**
```Python
return ErrorObservation(
    f'Fatal Runtime in editing: Expected FileWriteObservation, got {type(obs)}: {str(obs)}',
)
```

**Improved Code:**
```Python
return ErrorObservation(
    f'Error in editing: Expected FileWriteObservation, got {type(obs)}: {str(obs)}',
)
```

**Details:**
Consider using a consistent naming convention for error handling to improve readability and maintainability.

ngduyanhece commented on tests/unit/test_agent_controller.py:
```diff
@@ -87,7 +87,7 @@ async def test_on_event_change_agent_state_action(mock_agent, mock_event_stream)
 
 @pytest.mark.asyncio
-async def test_report_error(mock_agent, mock_event_stream):
+async def test__react_to_error(mock_agent, mock_event_stream):
```
**Suggestion:** Ensure test names reflect the functionality being tested.

**Existing Code:**
```Python
async def test_report_error(mock_agent, mock_event_stream):
```

**Improved Code:**
```Python
async def test_react_to_error(mock_agent, mock_event_stream):
```

**Details:**
Ensure that the test names clearly reflect the functionality being tested for better clarity.

xingyaoww commented on openhands/controller/state/state.py:
```diff
@@ -157,6 +154,12 @@ def __setstate__(self, state):
 
         # remove the restored data from the state if any
 
+    def get_last_error(self) -> str:
+        for event in self.history.get_events(reverse=True):
+            if isinstance(event, ErrorObservation):
+                return event.content
+        return ''
```
```suggestion
    def get_last_error(self) -> str:
        for event in self.history.get_events(reverse=True):
            if isinstance(event, ErrorObservation):
                return event.content if not event.fatal else 'There was a fatal error during agent execution: ' + event.content
        return ''
```

enyst replied:
There may be a problem here, illustrated by Xingyao's comment and https://github.com/All-Hands-AI/OpenHands/pull/4575/files#diff-cf5322e360ed33c3dc16023bfd443e6a7edab81c2f4a9e0dfd2f681a2fd59165R413

When client scripts like evals ask for `get_last_error()`, they need to know whether the agent terminated abnormally. Not what indentation blunder it did several steps before ending normally... 🤔

rbren replied:
This is a good point--fixed!

enyst commented on openhands/controller/agent_controller.py:
```diff
@@ -140,17 +138,15 @@ async def update_state_after_step(self):
-        await self.event_stream.async_add_event(
-            ErrorObservation(message), EventSource.AGENT
-        )
+        raise e
```
If this is all that's left of this method... it seems we don't really need it, maybe we can set the `agent_state` in the catch block 🤔
Is this state ever different than ERROR?

enyst replied:
There is one (one...) PAUSED

enyst commented on openhands/controller/agent_controller.py:
```diff
@@ -1,7 +1,7 @@
 import copy
 import traceback
-from typing import Type
+from typing import Callable, Optional, Type
```
```suggestion
from typing import Callable, Type
```

enyst commented on openhands/controller/agent_controller.py:
```diff
@@ -77,6 +75,7 @@ def __init__(
         initial_state: State | None = None,
         is_delegate: bool = False,
         headless_mode: bool = True,
+        status_callback: Optional[Callable] = None,
```
```suggestion
        status_callback: Callable | None = None,
```

enyst commented on openhands/runtime/impl/eventstream/eventstream_runtime.py:
```diff
@@ -499,17 +501,17 @@ def run_action(self, action: Action) -> Observation:
             except Exception as e:
                 self.log('error', f'Error during action execution: {e}')
-                obs = FatalErrorObservation(f'Action execution failed: {str(e)}')
+                obs = ErrorObservation(f'Action execution failed: {str(e)}')
```
Just setting a marker here... This and the one at line 504 are a tad suspicious, they're catch-alls and would go to the agent.

rbren replied:
Think I got these! But we really need to audit this--`except Exception` should rarely be used

<img width="354" alt="Screenshot 2024-10-31 at 8 54 20 AM" src="https://github.com/user-attachments/assets/03826275-3c7d-4ba9-82fc-4f3b97644837">

rbren replied:
Think I got these! But we really need to audit this--except Exception should rarely be used

![Uploading Screenshot 2024-10-31 at 8.56.53 AM.png…]()

rbren replied:
<img width="951" alt="Screenshot 2024-10-31 at 8 56 53 AM" src="https://github.com/user-attachments/assets/d010ccd6-5b51-4c60-a4b5-edb3894ffe32">

enyst replied:
OK. The first dozen+ are safe without looking, but I'll look anyway. 😂

enyst replied:
- not sure why are we re-raising [here](https://github.com/All-Hands-AI/OpenHands/pull/4575/files#diff-3f4657eb255ce8ef87dcec0e01591aa7e11600266ff7c94a1845cdbec0423e39R109); in save, it would result in not closing properly; in restore, it could be nicer. Not an issue for this PR IMO.

OK I started to look into these, I think we need to separate what is in scope because there might indeed be some cleanups we can make, but some may need some bit of discussion so IMO better in a follow-up.

Because I think we should get this PR in ASAP. I saw a "fatal" last night in the LLM prompt, and while I can take a hammer and knock it, ... we really need to clean this up.

enyst commented on openhands/runtime/impl/eventstream/eventstream_runtime.py:
```diff
@@ -500,17 +497,14 @@ def run_action(self, action: Action) -> Observation:
-                    obs = FatalErrorObservation(
-                        f'Action execution failed: {error_message}'
+                    obs = ErrorObservation(
+                        f'Action execution failed: {error_message}',
```
Just to note - this one is an ErrorObs for any non-200 status. Is it too broad? Are all those messages appropriate for the llm?

rbren replied:
ahh good catch. We actually raise for non-200, so this never gets hit. Removed

enyst commented:
Some notes:
- first run, I'm seeing a huge amount of "Getting container logs.."; might be from the logging PR or some combo with the timeouts changes, etc, here
- running with `main.py`: the first agent response is a MessageAction, not displayed in console
    - oh, the user enters something ("print 55")
    - then the _previous_ agent message is displayed in console ("please let me know what you'd like me to do...")
    - then the Observation
    - then the Action that triggered the Obs 😅 
 
I have no idea yet if it's in main, and the "environment" PR is also a possible suspect, but this is odd enough that I'll just write it down. @rbren

rbren commented:
fixed the `Getting container logs..` spam! looking into the message thing...

rbren commented:
OK the message thing was really weird! First heisenbug I've seen in a long time--adding `prints` fixed it!

Turned out it was because of directly calling `async_add_event` instead of letting the event stream run the handlers in a different async loop. Fixed!

enyst commented on openhands/runtime/base.py:
```diff
@@ -131,15 +159,28 @@ async def on_event(self, event: Event) -> None:
+                self.log('error', f'Problematic action: {str(event)}')
+                self.send_error_message(err_id, str(e))
+                self.close()
+                return
```
Just a quick question here. This doesn't has no obs, and doesn't re-raise (shouldn't it?), while the loop is waiting for an obs. I tried to replicate this manually somehow and I raised just in the `try` block:
- running with `main.py`
- it close()ed runtime
- then it looks like it hangs and that's it. Loop forever.

rbren replied:
Hmm this is a good point. Maybe it should reraise. Will try

enyst replied:
I don't think it works yet. First, this results in a "coro not awaited". Second, if the exception is not propagated downstream and doesn't stop the loop, then it would still have `_pending_action` set and I don't see why that will not loop endlessly...

rbren replied:
OK I refactored a bunch here, and now main.py and cli.py behave nicely in the presence of fatal exceptions. took some plumbing though--could definitely simplify further in a future PR

enyst commented:
The unit tests job got stuck twice, and it looks like the 3rd time it isn't better. May be not a fluke, it gets stuck in the [controller tests](https://github.com/All-Hands-AI/OpenHands/actions/runs/11630293059/job/32403187055?pr=4575#step:8:81).

rbren commented:
Whoops, lingering calls to async_add_event

tofarr reviewed (approved):
LGTM!

enyst reviewed (dismissed):
Thank you for taking this up, Robert. It's been broken for a while.

You will not believe this 😅, it's true though. The other day I wanted to look at the branch locally, couldn't remember the name so I tried 'rb/error' and surprisingly it worked, I had a `rb/error_handling` on my machine. `git log` was showing exactly _this_ kind of work setting up the error related arch. Except... the commits were from April. 😂

It took us 6 months to break it! Hey that's not even that bad, but we did manage to, eventually. 😅

I think this branch is obviously on the right track, but I am a bit concerned that it doesn't work yet on a couple of issues.

Maybe we could use the new integration workflows to test some things?

rbren commented:
@enyst any particular workflows you see breaking?

enyst commented on openhands/controller/agent_controller.py:
```diff
@@ -420,13 +392,8 @@ async def _step(self) -> None:
-                FatalErrorObservation('Agent got stuck in a loop'),
-                EventSource.ENVIRONMENT,
-            )
+            await self._react_to_exception(RuntimeError('Agent got stuck in a loop'))
```
I think moving this check at the start of the `_step` might mean that you can never restore a session that ended in a loop... I didn't try it on the latest version, but I did on others; and well, it either works to stop or it doesn't...

This is not the same if it's at the end; then it stops, and then, upon session restore, the LLM has another chance this time. (I'll test the exact behavior, it might be a little non-obvious if it stops in `main.py`)

On a side note, I stopped merging stuff in this branch, since I noticed it's probably making it harder for you. Sorry about that. On this call, I did think that when this branch works fine, it _can_ be at the end again. I could be wrong?

rbren replied:
Just did a test--session recovery seems to work just fine if it gets stuck in a loop

enyst commented:
> @enyst any particular workflows you see breaking?

Oh, I see the eval tasks really don't work at the moment. From the [existing integration tests](https://github.com/All-Hands-AI/OpenHands/blob/4446d3180f4a4dec7b1ef47f0a36d10be8659b15/.github/workflows/eval-runner.yml#L63), I also don't see any that would fail, because they're happy paths, and we need things going wrong.

What seems likely to fail:
- evals error report
  - evals retrying on some errors, like swe-bench linked above
- restoring session with `main.py` after stuck
- maybe restoring session with UI after stuck
- when the runtime has a serious issue/exception, with `main.py`, it blocks all.

neubig commented on evaluation/swe_bench/run_infer.py:
```diff
@@ -409,11 +409,11 @@ def process_instance(
-            and 'fatal error during agent execution' in state.last_error
+            state.get_last_error()
+            and 'fatal error during agent execution' in state.get_last_error()
             and 'stuck in a loop' not in state.last_error
```
should this be `get_last_error()`?

xingyaoww replied:
```suggestion
            and 'stuck in a loop' not in state.get_last_error()
```

neubig reviewed (commented):
I looked through and in general it lgtm. I'd be happy to be a stamp of approval if necessary, but I think that probably @enyst and @xingyaoww are more familiar overall, and this was an XXL PR so I didn't have time to carefully check everything involved.

xingyaoww commented on openhands/runtime/impl/eventstream/eventstream_runtime.py:
```diff
@@ -466,10 +456,14 @@ def run_action(self, action: Action) -> Observation:
             action_type = action.action  # type: ignore[attr-defined]
             if action_type not in ACTION_TYPE_TO_CLASS:
-                return FatalErrorObservation(f'Action {action_type} does not exist.')
+                return ErrorObservation(
```
I think we need to return 400 here? This should fail hard because it should be a code implementation error - wrong action type should never be sent here?

rbren replied:
This part is client-side,  but yeah I get your point

xingyaoww commented on openhands/runtime/impl/eventstream/eventstream_runtime.py:
```diff
@@ -484,33 +478,22 @@ def run_action(self, action: Action) -> Observation:
                 self.log('error', 'No response received within the timeout period.')
-                obs = FatalErrorObservation(
-                    f'Action execution timed out after {action.timeout} seconds.'
+                obs = ErrorObservation(
```
Same here - we probably should also fail hard here. Every occurrence of `FatalErrorObservation` should translate to an Exception IMO. Maybe we should create a `FatalRuntimeException`.

rbren replied:
If an action times out though, can't the agent re-run the action with a longer timeout?

xingyaoww replied:
We have timeout handled on the action_execution_server side - they are expected to return request within the given time period. 
For example: https://github.com/All-Hands-AI/OpenHands/blob/aab7fa483b7127be70f236a9da3df0c44f21b478/openhands/runtime/utils/bash.py#L176-L205


When they don't, it likely means the server dies (which happens a lot during eval) -- this is really a fatal error that should stops everything

xingyaoww replied:
And when the server dies, even if the agent keep execute commands, it will just get "timeout" repeatedly so we are just spending LLM credits without making any progresses..

rbren replied:
OK I get it now, thanks!

xingyaoww commented on openhands/runtime/impl/remote/remote_runtime.py:
```diff
@@ -361,49 +325,53 @@ def run_action(self, action: Action) -> Observation:
             if action_type not in ACTION_TYPE_TO_CLASS:
-                return FatalErrorObservation(
-                    f'[Runtime (ID={self.runtime_id})] Action {action_type} does not exist.'
+                return ErrorObservation(
```
ditto - we should raise exception here instead error observation

rbren replied:
fixed!

xingyaoww commented on openhands/runtime/impl/remote/remote_runtime.py:
```diff
@@ -361,49 +325,53 @@ def run_action(self, action: Action) -> Observation:
+                obs = observation_from_dict(output)
+                obs._cause = action.id  # type: ignore[attr-defined]
+            except requests.Timeout:
+                obs = ErrorObservation(
```
ditto

rbren replied:
fixed!

xingyaoww commented on openhands/runtime/utils/bash.py:
```diff
@@ -329,6 +329,6 @@ def run(self, action: CmdRunAction) -> CmdOutputObservation | FatalErrorObservat
         except UnicodeDecodeError as e:
-            return FatalErrorObservation(
-                f'Runtime bash execution failed: Command output could not be decoded as utf-8. {str(e)}'
+            return ErrorObservation(
```
ditto - this shouldn't be an ErrorObservation (not an agent error), but rather an exception

rbren replied:
This one I'm not sure about--stdout can totally output things other than unicode. This one seems recoverable

xingyaoww replied:
Let's keep this then! This was very uncommon irrc

xingyaoww reviewed (changes requested):
Most other stuff LGTM! The biggest issue I'm seeing is that we turn a lot of `FatalErrorObservation` back to `ErrorObservation`. I think we should create a `FatalRuntimeException` to properly handle these cases, instead of using `ErrorObservation` which will be send to the agent. In those fatal error cases, there's nothing the agent can do to fix it; rather *we* should fix the infra for it. 

Those fatal errors took me a while to spot them one-by-one -- Removing them / turn them into `ErrorObservation` will likely break our evaluation suite when running at scale (e.g., you don't know if your instance run failed because of "runtime error" or "LLM error", which make things very hard to debug).

rbren commented:
@xingyaoww "fatal" now means "shut everything down, the environment is in an irrecoverable state and the user needs to take action"

If e.g. an action times out, that's not fatal--the agent can try a different action, or increase the timeout (IIRC that's an option for the agent)

rbren commented:
OK I think I've addressed your comments @xingyaoww and @enyst! One last question for you though Xingyao

Thanks for the feedback--definitely helpful

xingyaoww commented on openhands/controller/agent_controller.py:
```diff
@@ -142,22 +142,16 @@ async def update_state_after_step(self):
+        self,
+        e: Exception,
+    ):
+        await self.set_agent_state_to(AgentState.ERROR)
```
Should we also set the message of exception to `self.state.last_error` here?

rbren replied:
So I decided not to do it here, because it potentially hides runtime errors if the caller doesn't implement this properly. It's set in `openhands/core/loop.py` instead

xingyaoww reviewed (commented):
Thanks! -- Mostly LGTM! Few smaller things!

amanape reviewed (approved):
FE LGTM

rbren commented on openhands/runtime/impl/remote/remote_runtime.py:
```diff
@@ -361,49 +325,47 @@ def run_action(self, action: Action) -> Observation:
             if not hasattr(self, action_type):
-                return FatalErrorObservation(
-                    f'[Runtime (ID={self.runtime_id})] Action {action_type} is not supported in the current runtime.'
+                return ErrorObservation(
```
I thought about this one, but went back on it. If we're in an alternative runtime, it seems reasonable to report this back to the agent so it can try another strategy. I'm not sure we can expect every Agent to handle ahead of time what the runtime capabilities are

rbren commented on openhands/runtime/impl/remote/remote_runtime.py:
```diff
@@ -268,72 +256,48 @@ def _parse_runtime_response(self, response: requests.Response):
-            # Retry 404 & 503 errors for the /alive endpoint
-            # because the runtime might just be starting up
-            # and have not registered the endpoint yet
-            retry_fns=[is_404_error, is_503_error],
```
These should 100% be gone with the new production-grade runtime-api. If we're still seeing them, we have big problems on the backend and should fix them there.

Can you try running an eval to validate this?

rbren commented:
OK I actually saw the docker cleanup issue as well--will work on that.

Re: UnicodeDecodeError, I pushed back on making that a hard error, but maybe I'm wrong? Where did that weird byte come from?

@enyst any other blockers from you?

rbren commented:
OK added a more graceful handling of an existing container!

@enyst looking at your unicode issue it seems like you might have pasted something funky into the prompt...

enyst commented:
Re: unicode. I don't know, I just yelled at the LLM, hmm... OK, let's ignore until it happens again. 😅 

I didn't get to see the implementation of the new loop, and I didn't get it stuck.

No blockers though.

xingyaoww commented:
@rbren I migrate all the core eval logics into a single testcase.
We can start it by doing:
```
export ALLHANDS_API_KEY="ah-YOUR-KEY"
export RUNTIME=remote
export SANDBOX_REMOTE_RUNTIME_API_URL="https://runtime.eval.all-hands.dev" 
pytest -vvxss tests/runtime/test_stress_remote_runtime.py
```

I changed number of workers to 1 in this PR, yet got this issue:
![image](https://github.com/user-attachments/assets/a156b21e-9d99-4712-aab2-6e3936b846e2)

rbren commented:
Thanks Xingyao, testing on my end!

@enyst you're still blocking merge on this one--if you're good w/ it, mind approving?

enyst commented:
Yes, in my opinion we should take this in as soon as we can. It's been getting difficult/slow to solve some types of issues on `main` without side effects, because the arch is broken and things are cropping up in the strangest places. We can work easier on this. ❤️

xingyaoww commented:
My last concern is just evaluation (e.g., we remove a bunch of 503/404 wait which will break eval :( -- but if we hope to get this in soon for the release, i can definitely unblock this.

xingyaoww reviewed (approved):
let me unblock first and let @rbren decide 🙏

rbren commented:
Stress tests are passing now! Thanks for writing that Xingyao, it did surface a bug where `/alive` returns 503 for a second even after the pod is ready (going to guess due to Traefik/ingress not being ready)

I think we're good to go! Enabling auto-merge...
